### PR TITLE
Hash-to-curve jets

### PR DESCRIPTION
**Pull request by roconnor-blockstream**
Add a `hash-to-curve` jet, compatible with the function that maps Asset IDs to Asset Commitments in Elements, and it's subroutine `shallue-van-de-woestijne`.

Note: some refactoring of jet specifications has changed the CMR of some secp256k1 jets.  Thus we are merging into a `staging` branch to keep master compatible with the current Simplicity testnet.

### CODE REVIEW

**Comment by apoelstra** [comment]
utACK e08df2cfec393b29ceb893639cbce670f1ea83e8. But my local CI fails with

```
gcc -c -DPRODUCTION -I include -Werror -Wall -Wextra -Wcast-qual -Wcast-align -Wstrict-aliasing -Wpointer-arith -Winit-self -Wshadow -Wswitch-enum -Wstrict-prototypes -Wmissing-prototypes -Wredundant-decls -Wfloat->
In file included from secp256k1/secp256k1_impl.h:68,
                 from jets-secp256k1.c:6:
secp256k1/generator_impl.h: In function 'secp256k1_generator_generate_internal':
secp256k1/generator_impl.h:218:39: error: declaration of 'secp256k1_scalar_zero' shadows a global declaration [8;;https://gcc.gnu.org/onlinedocs/gcc/Warning-Options.html#index-Wshadow-Werror=shadow8;;]
  218 |         static const secp256k1_scalar secp256k1_scalar_zero = SECP256K1_SCALAR_CONST(0, 0, 0, 0, 0, 0, 0, 0);
      |                                       ^~~~~~~~~~~~~~~~~~~~~
In file included from secp256k1/secp256k1_impl.h:18:
secp256k1/scalar_impl.h:26:31: note: shadowed declaration is here
   26 | static const secp256k1_scalar secp256k1_scalar_zero = SECP256K1_SCALAR_CONST(0, 0, 0, 0, 0, 0, 0, 0);
      |                               ^~~~~~~~~~~~~~~~~~~~~
cc1: all warnings being treated as errors
make: *** [Makefile:15: jets-secp256k1.o] Error 1
```

This seems like a reasonably straightforward thing to fix (though I'm not sure exactly what to do; maybe you need to #include scalar_impl.h directly in generator_impl.h, and remove your explicit definition of secp256k1_scalar_zero?), but I'm surprised that the real CI doesn't catch it.

**Comment by roconnor-blockstream** [comment]
Fixed that CI issue.

**Review by apoelstra — APPROVED** [review]
ACK 1142f14eac6eef08b729b96acf5b4ae2ff0bdf58

File: C/primitive/elements/primitiveJetNode.inc [hunk 1]
```diff
@@ -3471,6 +3479,14 @@
 , .targetIx = ty_pbw8
 , .cost = 149 /* milli weight units */
 }
+,[SWU] =
+{ .tag = JET
+, .jet = swu
+, .cmr = {{0x69459b6du, 0xd10a63dfu, 0x37dedf18u, 0x6cb4d208u, 0xd8c03961u, 0x1cecdd09u, 0xcfe07ac9u, 0x1d01422fu}}
+, .sourceIx = ty_w256
+, .targetIx = ty_w512
+, .cost = 10956 /* milli weight units */
```
[the remark this answers is not in the document]
Comment (roconnor-blockstream) [reply C/primitive/elements/primitiveJetNode.inc:3488]: yes.

**Review by uncomputable — APPROVED** [review]
ACK 1142f14eac6eef08b729b96acf5b4ae2ff0bdf58 C and Haskell tests run on my machine.

**Pull request merged**